### PR TITLE
Bug fix for sparse recorder

### PR DESCRIPTION
**Pull request by 123malin**
修复sparse recorder在full-async模式下造成的内存泄露问题，sparse-recorder仅在GEO模式下打开。

listen_and_serv op属性变动sync_mode(bool) 更改为distributed_mode(int)，原因及兼容性问题陈述如下：
1. 当前分布式训练模式有四种，GEO，full-async，half-async，sync，原有的属性sync_mode为bool类型，不够用。
2. listen_and_serv是分布式训练tranpiler内部调用的一个op，用户并不会直接接触和调用该OP，因此对它的变更不会对用户使用分布式训练的方式造成影响。
3. sync_mode属性仅用于server端区分分布式训练模式，对模型训练、保存、预测等均无影响，不会存在兼容问题。

因此从代码可维护性、简洁的角度出发，现决定不保留原有属性sync_mode(bool)，改用distributed_mode(int)。

### CODE REVIEW

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 1]
```diff
@@ -62,6 +62,13 @@
 PRINT_LOG = False
 
 
+class TrainingMode:
+    SYNC = 0
+    HALFASYNC = 1
```
Comment (seiriosPlus) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: HALFASYNC -> HALF_ASYNC

Comment (123malin) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: 嗯嗯 可以

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 2]
```diff
@@ -311,6 +318,13 @@ def __init__(self, config=None):
         if self.config.split_method is None:
             self.config.split_method = RoundRobin
 
+        if self.config.sync_mode:
+            self.training_mode = TrainingMode.SYNC
+        elif self.config.runtime_split_send_recv:
+            self.training_mode = TrainingMode.ASYNC
+        else:
+            self.training_mode = TrainingMode.HALF_ASYNC
```
Comment (seiriosPlus) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: 4种模式， 少判断一个~

**Review by seiriosPlus — DISMISSED** [review]
LGTM

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 2 above)
Comment (123malin) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: GEO在GeoSgdTranspiler.py文件里

File: paddle/fluid/operators/collective/c_gen_nccl_id_op.cc [hunk 3]
```diff
@@ -90,7 +90,7 @@ class CGenNCCLIdOp : public framework::OperatorBase {
     // NOTE: Can not use unique_ptr here because the default
     // deleter will call GRPC Server's base class's dtor and
     // that will cause a wired crash.
-    distributed::RequestSendHandler rpc_h(true);
+    distributed::RequestSendHandler rpc_h(distributed::TrainingMode::kSync);
```
Comment (guru4elephant) [inline paddle/fluid/operators/collective/c_gen_nccl_id_op.cc]: training mode can not generalize distributed prediction, currently I am ok with this.

**Review by guru4elephant — DISMISSED** [review]
LGTM, training mode should be generalized in distributed validation.

File: paddle/fluid/operators/collective/c_gen_nccl_id_op.cc (hunk 3 above)
Comment (123malin) [reply paddle/fluid/operators/collective/c_gen_nccl_id_op.cc]: This time, we need to solve the memory leak problem in async mode firstly. We'll continue to find a more generalized attribute name.

**Review by Shixiaowei02 — DISMISSED** [review]

**Comment by liym27** [comment]
LGTM for deleting`sync_mode` of OP  `listen_and_serv`.

**Review by liym27 — DISMISSED** [review]

**Comment by raindrops2sea** [comment]
> LGTM, training mode should be generalized in distributed validation.
How about parallel_mode / communication_mode/ sync_model  ? "training_mode" is too broad and  not easily associated with distributed strategy.

**Review by raindrops2sea — APPROVED** [review]
LGTM

**Review by Shixiaowei02 — APPROVED** [review]

**Review by seiriosPlus — APPROVED** [review]
LGTM

**Review by guru4elephant — APPROVED** [review]
LGTM

**Review by liym27 — APPROVED** [review]
